sdk/core/azure_core_amqp/src/connection.rs

```diff
@@ -0,0 +1,315 @@
+    fn open(
+        &self,
+        name: impl Into<String>,
+        url: Url,
+        options: Option<AmqpConnectionOptions>,
+    ) -> impl std::future::Future<Output = Result<()>> {
+        self.0 .0.open(name, url, options)
+    }
+    fn close(&self) -> impl std::future::Future<Output = Result<()>> {
+        self.0 .0.close()
+    }
+}
+
+impl AmqpConnection {
+    pub fn new() -> Self {
+        let connection = ConnectionImplementation::new();
+
+        Self(AmqpConnectionImpl::new(connection))
+    }
+}
+
+pub mod builders {
+    use super::*;
+    pub struct AmqpConnectionOptionsBuilder {
+        options: AmqpConnectionOptions,
+    }
+
+    impl AmqpConnectionOptionsBuilder {
+        pub(super) fn new() -> Self {
+            Self {
+                options: AmqpConnectionOptions {
+                    max_frame_size: None,
+                    channel_max: None,
+                    idle_timeout: None,
+                    outgoing_locales: None,
+                    incoming_locales: None,
+                    offered_capabilities: None,
+                    desired_capabilities: None,
+                    properties: None,
+                    buffer_size: None,
+                },
+            }
+        }
+        pub fn build(self) -> AmqpConnectionOptions {
+            self.options
+        }
+        #[allow(dead_code)]
+        pub fn with_max_frame_size(mut self, max_frame_size: u32) -> Self {
+            self.options.max_frame_size = Some(max_frame_size);
+            self
+        }
+        #[allow(dead_code)]
+        pub fn with_channel_max(mut self, channel_max: u16) -> Self {
+            self.options.channel_max = Some(channel_max);
+            self
+        }
+        #[allow(dead_code)]
+        pub fn with_idle_timeout(mut self, idle_timeout: Duration) -> Self {
+            self.options.idle_timeout = Some(idle_timeout);
+            self
+        }
+        #[allow(dead_code)]
+        pub fn with_outgoing_locales(mut self, outgoing_locales: Vec<String>) -> Self {
+            self.options.outgoing_locales = Some(outgoing_locales);
+            self
+        }
+        #[allow(dead_code)]
+        pub fn with_incoming_locales(mut self, incoming_locales: Vec<String>) -> Self {
+            self.options.incoming_locales = Some(incoming_locales);
+            self
+        }
+        #[allow(dead_code)]
+        pub fn with_offered_capabilities(mut self, offered_capabilities: Vec<AmqpSymbol>) -> Self {
+            self.options.offered_capabilities = Some(offered_capabilities);
+            self
+        }
+        #[allow(dead_code)]
+        pub fn with_desired_capabilities(mut self, desired_capabilities: Vec<AmqpSymbol>) -> Self {
+            self.options.desired_capabilities = Some(desired_capabilities);
+            self
+        }
+        pub fn with_properties(
+            mut self,
+            properties: Vec<(impl Into<AmqpSymbol>, impl Into<AmqpValue>)>,
```

Comment:
Is a tuple an expected form that the user will pass? Just wondering if all this type coercion is actually necessary, especially if it's a corner case. We can always add helpers eventually but should probably keep it lower level for now.

Reply:
Actually, this one should be "impl Into<AmqpOrderedMap<AmqpSymbol, AmqpValue>>" which is more accurate.